### PR TITLE
AFK subsystem now also despawns certain antags. message_admins replaced with log_admins. Better feedback on enabling

### PR DESCRIPTION
**Pull request by farie82**
Id like a test merge for this PR prior of it getting merged

## What Does This PR Do
1. Makes the system able to cryo/despawn certain antags. Excluding wizards, slings, abductors and nukies. Roles like slaughter demon should not even be candidates to the system
2. The toggle in the preferences now tells you what happens if you switch it on
3. The system now will make less use of message_admin and instead just log it.

## Why It's Good For The Game
Seems there are still issues with the current system so I hope this addresses those.

## Images of changes
image
(1 minute in the cryostorage message is for debugging. Usually is 20)

image
Only one message for admins if the system can't despawn certain players

## Changelog
:cl:
tweak: The autocryoing preferrence now also works for certain antags. Others like shadowlings won't get cryod but instead will send a message to admins that they are AFK.
tweak: The autocryoing preference now will give feedback in the chat informing the player about the option when used.
tweak: SSAfk now sends less messages to admins. Only when it cannot make a decision itself.
/:cl:

### CODE REVIEW

File: code/modules/client/preference/preferences.dm [hunk 1]
```diff
@@ -2000,7 +2003,22 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
...
 				if("afk_watch")
-					afk_watch = !afk_watch
+					var/afk_new = input(user, "Select a new amount of minutes after which you get warned for being AFK, between 0 and 10 (0 will deactivate)", UI_style_alpha) as num
+					if(afk_new == 0)
+						// deactivate the system
+						AFK_WATCH_warn_minutes = 0
+						AFK_WATCH_cryo_minutes = 0
+						AFK_WATCH_antag = FALSE
+					else 
+						if(!afk_new | !(afk_new <= 10 && afk_new >= 1)) return
```
Comment (farie82) [inline code/modules/client/preference/preferences.dm]: Need values for the boundries here from the heads/maints

File: code/modules/client/preference/preferences.dm [hunk 2]
```diff
@@ -2000,7 +2003,22 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
...
+						AFK_WATCH_warn_minutes = afk_new
+						afk_new = input(user, "Select a new amount of minutes after which you get cryod for being AFK, between 1 and 5", UI_style_alpha) as num
+						if(!afk_new | !(afk_new <= 5 && afk_new >= 1))
```
Comment (farie82) [inline code/modules/client/preference/preferences.dm]: Same here. Boundries required

File: code/controllers/subsystem/afk.dm [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+#define AFK_FAST_CRYO_WARN 3
+#define AFK_FAST_CRYO_CRYO 5
 
 SUBSYSTEM_DEF(afk)
 	name = "AFK Watcher"
 	wait = 300
 	flags = SS_BACKGROUND
 	var/list/afk_players = list() // Associative list. ckey as key and AFK state as value
+	var/list/non_cryo_antags = list( \
+								SPECIAL_ROLE_ABDUCTOR_AGENT, SPECIAL_ROLE_ABDUCTOR_SCIENTIST, \
```
Comment (farie82) [inline code/controllers/subsystem/afk.dm]: Weird spacing on gits part. Locally it looks fine

**Comment by farie82** [comment]
Updated the PR after talking with Kyet about it.
Changes:
1. Added player timing preferences are gone. It will be 2 sets of timers now. The default timer for all players (Kyet said it currently is 20 minutes after you get cryod) and the fast cryo method (activated in a players preference) which is 5 minutes till you get cryod.
2. Certain antags now can get cryod. For others a message will be send to the admins so that they can offer the body to ghosts or something

File: code/controllers/subsystem/afk.dm [hunk 4]
```diff
@@ -1,11 +1,18 @@
-#define AFK_WARNED       1
-#define AFK_CRYOD    	 2
+#define AFK_WARNED        1
+#define AFK_CRYOD    	  2
+#define AFK_ADMINS_WARNED 2
```
Comment (farie82) [inline code/controllers/subsystem/afk.dm]: Same value here as AFK_CRYOD since the behaviour will be the same for now

**Comment by Pixelzilla** [comment]
> 
> 
> Updated the PR after talking with Kyet about it.
> Changes:
> 
>     1. Added player timing preferences are gone. It will be 2 sets of timers now. The default timer for all players (Kyet said it currently is 20 minutes after you get cryod) and the fast cryo method (activated in a players preference) which is 5 minutes till you get cryod.
> 
>     2. Certain antags now can get cryod. For others a message will be send to the admins so that they can offer the body to ghosts or something

Does this mean that players will have absolutely no way to disable this feature now?

**Comment by farie82** [comment]
Aye it'll be active for everybody then. I trust the heads to keep the times to a generous number so that nearly nobody notices it and only real AFKers get cryod.

**Comment by Pixelzilla** [comment]
> 
> 
> Aye it'll be active for everybody then. I trust the heads to keep the times to a generous number so that nearly nobody notices it and only real AFKers get cryod.

Gonna have to say no support. I dont think this is a feature that really should exist in the first place. It already breaks immersion to see SSD people automagically teleport to be removed. Now we're gonna see **more** of that from players who haven't actually closed their clients but just dared to be idle for "too long"? What if they're reading the wiki? What if they're making food? They've not closed the game. Isn't the goal to keep players in the game instead of forcing them to wait for a new round or a tiny chance of rejoining (as something else)?

Automation is fine for some things but I dont think an afk system is one of them. Someone being afk for a little doesn't always impact things in a problematic way. Sure, if they're a head or an antag but those are cases that should fall to admins doing something about it (thats why admins exist, because people can actually evaluate a situation whereas an automatic system cannot).

**Comment by Kyep** [comment]
@Pixelzilla 
When a character is teleported to cryo, it should spawn a brief bluespace portal on them, named so as to tell observers it is part of NT's station system for handling crew that fall SSD. So, ICly, there is an explanation. That's how the SSD-handling system works, and this should work the same way.
The setting for this is that they're moved to cryo after 20 minutes, and despawned after 21 minutes. That's quite awhile to read the wiki / make food, and the timer resets if they do literally anything during that time. Even move one inch. Not to mention they get a warning (with sound effect) several minutes before it moves them, so if they're paying any attention at all, they can take a step or something to reset the timer.
If a player is playing the game, yes, we want to keep them in the game, but unfortunately there are players who will go AFK for literally an hour or more... which can be a real problem. Command staff going AFK that long essentially leaves their department without oversight most of the round. Antags going AFK that long leave themselves vulnerable to being looted. Certain antags (like cultists) going AFK that long can get them pickpocketed, not only crippling their team by getting one of their few team members caught, but also outing the round type. Not to mention all the standard issues of AFK players taking job slots that could be used by active players, and AFK players generating ahelps like "so and so has been afk for 30+ minutes, and I need X".

If someone has even the slightest inclination to actually play the game, sure, keep them around. But if they're AFKing for 20+ minutes despite both visual and audio warning that they'll be despawned... they probably need to be moved to cryo.

When SSDs, for example, were handled manually by admins, admins spent a good chunk of their time just managing SSDs, which was tedious, unnecessary, and generally a waste of admin time. When we added the SSD management system to move SSD people to cryo after awhile, it freed up a substantial amount of admin time to do more useful things, not to mention reducing the number of cases of people attacking SSDs and otherwise creating problems.
Similarly, there's no reason to force admins to handle AFK players manually when a perfectly good system for automatically resolving these cases is available. Yeah, it will mean that players can't just go AFK for an hour anymore and have their character be a statue for that time - or worse come back and find themselves looted and require a bunch of admin intervention to fix - but the fact we'll be automatically avoiding this is a good thing.
Most games are actually far less generous than I intend us to be - the other games I normally play will disconnect you if you're AFK for about 3-5 minutes, whereas we allow over 20. 

Oh, and I should also mention, this system still allows admins to override it in exceptional cases. If someone has a RL emergency, comes back 45 minutes later, and finds they were cryoed and ghosted, they can just ahelp, asking the admins to move them back to lobby so they can rejoin - we can do that with a button in their player panel.
Better that we handle the rare, exceptional cases like that manually - than having to handle EVERY CASE manually. An automated system like this doesn't remove the scope for admins to evaluate cases and make a decision - it just frees the admins up by handling cases automatically by default, and letting the admins intervene in exceptional cases if necessary, rather than forcing the admins to tediously waste time handling every single case manually no matter how clear cut it is. People can and do AFK, even in critical roles, for over an hour, on a regular basis. Those people need to be cryoed so those roles can be filled by someone who is active.

**Comment by Pixelzilla** [comment]
@kyep
I understand how the system works? I just think it breaks immersion for it to work that way. The portal effect is also used elsewhere in the game so is easily mistaken for that. Is someone teleporting around the station or is it just an SSD person? Who knows.

I've played a very very long time and I dont think the issues with people being ssd has actually been that big of a problem because players can and often would drag ssds to cryo. 

I can overlook and even accept the system for SSD people even if I personally think it is a lazy solution that disrupts RP with a bit of magical handwaving. What I absolutely do not support is it being used when a player is afk. This system being forced on everyone with no way for the player to opt out is honestly really disappointing.

If someone is afk they are not and have never been protected by looting/ssd rules. Thats the risk they choose (or maybe had to) take. There are plenty of valid reasons someone might need to go afk for 20+ minutes but still wish to return to the game and those players shouldn't have to depend on admins to be on/willing/not busy to essentially respawn them. Saying you can just boot someone from deadchat and respawn them is....weird because then you're giving someone a bunch of intel from ghost chat and probably increasing admin workload more than it needs to be.

Instead of spending a fraction of time contacting afk heads/antags and dealing with that you'll now get to spend a bunch more time dealing with players who got removed from the round and are upset/frustrated/annoyed because they have no way to opt out and have to depend on admins to decide to respawn them.

 Should admins handle every player? Probably not. Can admins focus on critical roles and antags? Absolutely. I'm sure players will be the ones ahelping to inform admins about critical roles that are afk for too long. But there absolutely is a reason for admins to handle important afk roles and thats because admins are not an automated system and are capable of seeing the bigger picture and the context of the situation where this system treats everyone in an equally dismissive way. 

You cant compare ss13 to other games. Most other games with anti-afking are ones where you can jump back in automatically to continue playing where you left off or in another match. This isn't really like any other games and I dont think it is fair to compare it to them. Games with actual roleplaying elements generally do strive to not break immersion and do generally require heavy administration - thats the job admins sign up for, the workload can be intense but that is a choice to take that on.

Removing the opt-in feature is a terrible idea for the players.
Instead of having admins having to occasionally deal with contacting a player in a critical role who's afk you're just going to be having a surplus of players contacting you wanting to be put back in the game and that's just going to open up the door for a lot of meta from ghost chat and a lot of frustration since being able to rejoin doesn't exactly give that player any of the things they had back.

**Comment by Fox-McCloud** [comment]
I'm opposed to any system that fully automates the AFK system.

We had this in the past. It was frustrating and accomplished little, in the long term.

In the day and age of being able to easily open job slots, by the HoP and admins alike, this is overkill.

**Comment by Kyep** [comment]
@Fox-McCloud I realize you may not like it, but IMHO it is necessary. We've had issues with entire rounds being spoiled by the lack of such a system. E.g. one cultist goes afk, gets searched, and suddenly the whole station goes on alert for cult because one AFK guy spoiled the round type. Plus, having a system like this enables people with certain medical issues to play roles like command safe in the knowledge that if they drop AFK they'll be quickly despawned (this is optional).
Your argument that it was frustrating hasn't been proven to be true - the system already has a default auto-cryo for people AFK for 20m in most situations. And we get essentially zero complaints about that. This PR is about cleaning up problematic edge cases and being consistent, not really automation.

**Comment by AffectedArc07** [comment]
I dont like this. I quite often have to go AFK for random chunks of time (hence why I usually play Civilian roles), and its usually fine to just come back. If I had to leave near the start of a round and then come back to being auto-cryo'd, I wont be happy that I cant play the game for another 2 hours.

I would however, be fine with auto-moving to the cryopods, but without the full despawning.

**Comment by Kyep** [comment]
@AffectedArc07 

- The system already moves you to cryo if you are AFK for 20+ minutes. Unless you specifically turn that option off. Have you turned it off? You get a warning after AFKing for 15 minutes that should (if memory serves) play a sound, so its easy to avoid being cryoed for AFKing even if you AFK a long time so long as you respond to the alert you get.
- Even if it did affect you and you were cryoed, it would be easy to ahelp 'hey I had to go afk, please send me back to lobby'. Its easier for admins to deal with that in the very rare case it happens, than to deal with people going afk regularly (often not playing civilian, but more significant jobs).
- If someone is in a cryopod, they should be despawned after a bit. Holding them in cryo forever without despawning them is bad.
- We've been running the AFK->cryo system in its default ON state for awhile now and I have yet to see even a single person in-round ahelp about it.

**Comment by AffectedArc07** [comment]
I explicitly turned that system off in preferences. I guess ahelping to be sent back to lobby works, but at the same time thats hassle for the admins who are already busy enough.

File: code/modules/client/preference/preferences.dm [hunk 5]
```diff
@@ -2000,8 +2000,10 @@ GLOBAL_LIST_INIT(special_role_times, list( //minimum age (in days) for accounts
 				if("winflash")
 					windowflashing = !windowflashing
 
-				if("afk_fast_cryo")
-					afk_fast_cryo = alert(user, "Would you like to be qualified for the fast cryoing? Being AFK for 5 minutes will result in cryoing then (instead of [config.auto_cryo_afk])", "Fast cryoing", "Yes", "No") == "Yes"
+				if("afk_watch")
+					to_chat(user, "<span class='info'>You will now get put into cryo dorms after [config.auto_cryo_afk] minutes. \
+								Then after [config.auto_despawn_afk] minutes you will be fully despawned. You will receive a visual and auditory warning before you will be put into cryodorms.</span>")
+					afk_watch = !afk_watch
```
Comment (AffectedArc07) [inline code/modules/client/preference/preferences.dm]: Correct me if I am wrong, but this is going to show up whether you turn this thing on or off, making is a huge confusion

Comment (farie82) [reply code/modules/client/preference/preferences.dm]: Aye. Saw it when I pushed the changes. Updated now

**Comment by farie82** [comment]
The PR has been changed. 
It is not forced upon everybody again. It still is an option now that you can turn on or off.
The fast despawning part is removed to favor no config or DB change

File: code/controllers/subsystem/afk.dm [hunk 6]
```diff
@@ -67,9 +79,8 @@ SUBSYSTEM_DEF(afk)
 	if(H.client)
 		window_flash(H.client)
 
-/datum/controller/subsystem/afk/proc/msg_admins(mob/living/carbon/human/H, mins_afk,  turf/location, action, info)
+/datum/controller/subsystem/afk/proc/log_admins(mob/living/carbon/human/H, mins_afk, turf/location, action, info)
```
Comment (AffectedArc07) [inline code/controllers/subsystem/afk.dm:82]: Just a nitpick but I hate how closely this proc is named to `log_admin`, and it seems ripe for confusion if anyone looks at this down the road

Comment (farie82) [reply code/controllers/subsystem/afk.dm]: Good point. Changed the name to `log_afk_action`

**Pull request merged**